tests/integration/service_test.py

```diff
@@ -161,6 +162,36 @@ def test_build_extra_hosts(self):
             {'www.example.com': '192.168.0.17',
              'api.example.com': '192.168.0.18'})
 
+    def test_build_ulimits(self):
+        def sort_ulimit_conf(ulimit_conf):
+            return sorted(ulimit_conf, key=lambda k: k['Name'])
+        # invalid options
```

Comment:
There's a bit too much going on in this test, can you split these out into separate tests please?  `test_build_ulimits_with_invalid_options`, `test_build_ulimits_with_dicts`, `test_build_ulimits_with_integers`